### DISCUSSION TITLE
Different synths on one ESP32 -- possible?

### DISCUSSION THREAD

**Discussion by copych** [marcel-licence/esp32_midi_sampler#89] [General]
Dear @marcel-licence , thank you one more time for the ability of implementing so many cool devices with just an ESP32 and a DAC! Doing this way I'd like to ask, basing on your experience, is it possible to have 2-4 moog or tb303-like monophonic instruments along with 10-12 voices of a sampler-based synth on same one ESP32? Maybe on different cores, but midi-manageable with one midi-in. While all the controls, ports and indications be held on some other board -- atmega, or another esp or whatever?

**Comment by copych** [comment]
@marcel-licence, may i reformat the question: is it   hard to make a synth combining different types of synthesis on one core, say, just one voice of a "simple synth" and one voice of a "sampler"?

**Reply by marcel-licence** [reply]
@copych I think that shouldn't be so difficult because the simple synth uses one sample for its waveform. You could integrate a simple waveform into the sampler (loop the complete sample) containing only one waveform cycle.

**Reply by copych** [reply]
@marcel-licence thank you for the reply, then the next question(s). I am not that good at programming, so sorry if you find it lame. is it possible to run 2 independent synth threads on 2 esp32 cores, they would prepare short buffers of synth data, and then main task sums these buffers and pushes to i2s? Or core0 is running synthesis and Core1 is processing effects chain?

**Reply by marcel-licence** [reply]
@copych in general it should be possible to use both cores for sound processing. It would increase the latency because the processing has to be done in sequence. But it would be possible to let them run independently. I do not know if the ESP32 or Arduino pushes tasks to the second core. At the moment I decided to run only one core with the time critical task and the second one for components which may need more time to be processed and which cannot be done at audio rate.

For example:
Core 1 will process the synth core and pushes its output to Buffer 1
Core 2 processes Buffer 2 at this time and pushes it to I2S when finished

Then the next buffer cycle starts:
Core 1 will process the synth core and writes the samples to Buffer 2 
Core 2 processes Buffer 1 and push the samples to I2S when finished.

The problem here FX can be only applied when the synth part is ready with its processing.

**Reply by copych** [reply]
@marcel-licence yes, i understand this, but what kind of delay will it be? 64 samples at 44100Hz is 1.4ms which is nothing, as well as 2.8ms. But actually I don't know what buffer size is used in your SynthTools.
BTW aren't you going to add a moog-like synth in your library?

**Comment by copych** [comment]
@marcel-licence sorry for bothering, but I couldn't help testing, and here's the POC. https://github.com/copych/ESP32-dualcore-i2s/tree/main

**Reply by marcel-licence** [reply]
@copych thank you. Looks like a good example. I hope I can test it soon. For a headless application it would be nice to use both cores for audio processing. At the end it would be possible to split the polyphony to both cores or process the audio in series. 
Regarding the moog-like synth. Yes I have a plan I want to implement in future. It might be limited to a reduced polyphony with a better audio sound quality in focus.
